src/scripts/blah_common_submit_functions.sh

```diff
@@ -633,7 +633,7 @@ function bls_start_job_wrapper ()
   fi
 
   echo "mkdir \$new_home"
-  echo "trap 'wait \$job_pid; cd \$old_home; rm -rf \$new_home; exit 255' 1 2 3 15 24"
+  echo "trap 'kill -s SIGQUIT \$job_pid; wait \$job_pid; cd \$old_home; rm -rf \$new_home; exit 255' 1 2 3 15 24"
```

Comment:
This turns the various signals we're trapping all into SIGQUIT. Can we propagate the other signals as well?

Reply:
>propagate the other signals

Marco's implementation looks fine for that, unless there's some way to get the caught signal number directly from within the trap.

My thoughts:

- can we use signal names (`HUP`, `INT`, `QUIT`, `TERM`, `SIGXCPU`(?), and `EXIT`) instead of numbers?

- is `rm -rf \$new_home;` safe?  is `$new_home` guaranteed to be free of whitespace and shell metacharacters?

- is bash running a foreground job when it is expected to receive a signal?  (vs a background job + `wait`)
  if so, i think the trap cmd will not get run until the foreground job completes

Reply:
- I don't see an issue with using signal names. 
- I believe we create `$new_home` but I don't remember how off-hand
- I'm pretty sure that this /is/ the foreground job

Reply:
> I believe we create `$new_home` but I don't remember how off-hand

Yeah [it looks like](https://github.com/opensciencegrid/BLAH/blob/7f5b6c9434886f96c729051ff99058e6866152d1/src/scripts/blah_common_submit_functions.sh#L622-L635) we create it, but the path is either under `$blah_wn_temporary_home_dir` (if defined in `blah.config`), or else `$old_home`, which is apparently just the working dir.

Anyway point is just that it's apparently controlled by stuff outside the script so `$new_home` should probably be quoted, especially if you're gonna do something as bold as `rm -rf` it.

> I'm pretty sure that this /is/ the foreground job

Right so, the question is, within this running script, which is a bash process which may run other commands, will the bash process be running the job in its foreground, or in its background with a `wait` command.  (If in the foreground, traps won't fire till the job command returns, but if waiting for a background job, the trap will fire right away.)

[Looks like](https://github.com/opensciencegrid/BLAH/blob/7f5b6c9434886f96c729051ff99058e6866152d1/src/scripts/blah_common_submit_functions.sh#L690) it waits for a background job though, so the trap should work as expected.

Reply:
I'd suggest quoting the directory name too. Then you cancel different directories upon normal termination (actually seems only the credentials). Any reason for that?
I looked about getting the signal number inside trap. someone mentioned online "Within the trap (when triggered via a signal), the $? variable is initially set to the signal number plus 128" but that did not work for me. Let me know if you find a way to get the signal number.
You may want to clear job_pid= after the wait for the job in the regular termination (or clear the traps)

Reply:
>someone mentioned online "Within the trap (when triggered via a signal), the $? variable is initially set to the signal number plus 128" but that did not work for me

As far as i can tell this is an undocumented myth.  If the shell was running a foreground job that got killed by a signal, then yes `$?` will be set, but that's because of the way the last command failed (the foreground job), not actually because of the signal that triggered the trap.  (Even if they happen to be the same).

Anyway unfortunately it doesn't seem to be a reliable way to get the signal number from within the trap, and i haven't found any documented way to do it either.  So your solution to set the handlers one at a time in a loop may be the best we can do.